include/zephyr/net/net_l2_connectivity.h

```diff
@@ -0,0 +1,120 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public API for managing L2 layers that have separated post-admin-up connect
+ *	  or associate phases
+ */
+
+#ifndef ZEPHYR_INCLUDE_NET_L2_CONNECTIVITY_H_
+#define ZEPHYR_INCLUDE_NET_L2_CONNECTIVITY_H_
+
+#include <zephyr/device.h>
+#include <zephyr/net/net_if.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
+/**
+ * @brief Network Layer 2 connectivity management abstraction layer
+ * @defgroup net_l2_connectivity Network L2 Connectivity Abstraction Layer
+ * @ingroup networking
+ * @{
+ */
+
+/**
+ * @brief Network L2 connectivity API structure
+ *
+ * Used to provide an interface to connect parameters and procedures
+ */
+struct net_l2_connectivity_api {
+	int (*connect)(struct net_if *iface);
+	int (*disconnect)(struct net_if *iface);
+
+	/* Note to reviewers:
+	 * Technically, as long as each iface provides a means by which to set
+	 * association/connection parameters, these get/set_opt funcs are not needed.
+	 *
+	 * However, it can be argued that they should still exist, since they establish a
+	 * standardized pattern for doing that.
+	 *
+	 * The question, then, is: Why limit these to specifically connectivity?
+	 * Should these apply to the whole iface, and be distributed to implementations
+	 * (where they are present) on not only the connectivity_api, but also the l2_api
+	 * and if_api as well?
+	 *
+	 * For now, I have implemented them just on the connectivity_api, but I wonder if they
+	 * should perhaps be more generalized.
+	 */
+	int (*set_opt)(int optname, const void *optval, size_t optlen);
+	int (*get_opt)(int optname, const void *optval, size_t *optlen);
+};
+
+/**
+ * @brief Network interface connectivity structure
+ *
+ * Binds L2 connectivity API and settings to an iface / network device
+ */
+struct net_if_conn {
+	/* The network device the connectivity struct is bound to */
+	struct net_if_dev *if_dev;
+
+	/* The L2 connectivity implementation associated with the network device */
+	struct net_l2_connectivity_api *api;
+
+	/* Per-net-device connectivity settings */
+	int timeout;
+	bool persistence;
+
+	/* To reviewers: Perhaps connectivity settings should be per-l2, rather than per-iface?
+	 *
+	 * If so, we could rename net_l2_connectivity_api to just net_l2_conn, and then move
+	 * the settings values into that.
+	 *
+	 * I actually think this would be an overall tidier solution. The net_if_conn struct would
+	 * become purely a binding between iface and net_l2_conn.
+	 */
+};
+
+/** @cond INTERNAL_HIDDEN */
+#define NET_IF_CONN_GET_NAME(dev_id, sfx) __net_if_conn_##dev_id##_##sfx
+/** @endcond */
+
+/**
+ * @brief Associate an L2 connectivity implementation with an existing network device instance
+ *
+ * @param dev_id Network device id.
+ * @param inst Network device instance.
+ * @param conn_api Pointer to net_l2_connectivity_api struct to associate.
+ */
+#define NET_DEVICE_INSTANCE_DEFINE_L2_CONNECTIVITY(dev_id, inst, conn_api)	\
```

Comment:
I wonder if we could follow the L2 approach here, i. e. instead of passing a pointer to the connectivity API struct, implement something similar to `NET_L2_INIT()` and `NET_L2_DECLARE_PUBLIC()` where specific connectivity implementation could register their existence and then be referenced here by name.

Reply:
If we go with this direction, i.e., `NET_L2_CONN_INIT()` and `NET_L2_DECLARE_PUBLIC()` then I would probably be more glad to see `NET_DEVICE_INIT()` macros modified to take the name of L2 CONN just as it takes L2 name. At least to me this would make it more logical compared to an init in a separate place.

While pitching this to Robert he pointed out that we should define the L2 CONN structures only if needed. I think that `COND_CODE_0` macro could help. How, I think, this could work is that if you don't use a L2 CONN then you pass a something that expands to 0 (NULL?). This in turn places empty block of code.

On the other hand if you pass a valid L2 CONN name, e.g., L2_CONN_WIFI then this would expand to `NET_DEVICE_INSTANCE_DEFINE_L2_CONNECTIVITY(..., ..., L2_CONN_WIFI)`;

Reply:
> While pitching this to Robert he pointed out that we should define the L2 CONN structures only if needed. I think that COND_CODE_0 macro could help. How, I think, this could work is that if you don't use a L2 CONN then you pass a something that expands to 0 (NULL?) to the modified NET_DEVICE_INIT() macro. This in turn places an empty block of code.

Yeah, were we to go this route, we'd just have callers of `Z_NET_DEVICE_INIT` or any of its wrappers/equivalents pass NULL for the l2_conn argument on interfaces that don't need a connectivity implementation

But I hesitate to add an argument to `NET_DEVICE_INIT`. Firstly, there are a total of `18` associated macros in Zephyr that would also have to be modified to accept this additional argument. And, in the Zephyr codebase alone, these macros are referenced a total of `172` times in `92` files, and the majority of these references would be simply passing NULL to the new argument.

Basically, I worry that further complicating this already pretty complex call tree may have long term detrimental effects on maintainability.

I'm willing to do it if there is strong agreement that it really is the way to go, but I just want to lift up that caution before doing so

Reply:
Yeah, I understand. And then I don't expect the L2 connectivity implementations to be added often, so I think what you propose is fine.